### DISCUSSION TITLE
[FEATURE REQ] Minimum Free Space

### DISCUSSION THREAD

**Discussion by CarlitoGil** [autobrr/autobrr#1652] [Ideas]
Could you please add a rule to each filter that checks if a minimum amount of free disk space is available?

I'd like to make sure that an approximate X amount of free space is always available for other filters that may need it later, or just make sure that there's enough space for things I actually want to watch.

**Comment by ghost** [comment]
This could potentially be a bit tricky to handle. In the mean time, I would suggest that you use a pre-filter script in this case.

**Comment by CarlitoGil** [comment]
Yeah, I supposed.
I can see that to make it perfect it would have to determine for sure the size of the torrent to be downloaded and perhaps see how many bytes remain to be downloaded in active torrents.
But, I'd be happy with just checking the free space on a specific drive at the time of grabbing the torrent.
Guess I can whip out the old VBScript skills.

**Comment by ghost** [comment]
Check out https://github.com/autobrr/autobrr/blob/develop/internal/domain/macros.go#L12
Docs are a bit outdated and can be found here: https://autobrr.com/filters/actions#macros

So you have two options:
1) send {{ .TorrentDataRawBytes }} to your script, parse torrent and calculate final size, then proceed on with your logic
2) see what {{ .Size }} represents (whether .torrent size or the total size of torrent files) -- I forgot to be honest and then proceed on with your logic

For the "how many bytes remain to be downloaded" you could use the "pre-allocate disk space" setting if your download client supports it, that way you take this factor out of the equation. Otherwise if you're on qBt (don't know about other clients), querying the API shouldn't be hard at all.

**Comment by zze0s** [comment]
Checking for space only works on the local system and one core goal with autobrr is to work the same way local and remote.

I've built a small http service that can run on remote systems to do these checks as an **External filter** that checks the status. I'll see about making it public.

**Comment by KyleSanderson** [comment]
> Checking for space only works on the local system and one core goal with autobrr is to work the same way local and remote.

I think there's an easier way to do this. the client itself has a "free space" query - using that should result in the expected behaviour.

**Comment by gennaios** [comment]
I haven’t used the script function so I’m not sure. Is such run after all other filters as a last step or does it execute to filter each announce? Checking free space or other I’d like to check only as needed if all other filters pass. Just verifying before I try creating a script.

**Comment by zze0s** [comment]
Copying from https://github.com/autobrr/autobrr/issues/630 and closing that as dupe

---

This is something that has been planned since the start of this project, but haven't made it's way into it yet because I want a solution that works with remote data etc.

Some clients do expose an endpoint to get free space, but not all do.

There is however a pretty simple solution that has been possible since we added the "external filters" that can run scripts.

## Freespace script

https://discord.com/channels/881212911849209957/938041606232211488/1025118932429963286

Something like this. Create the file somewhere `touch ~/freespace.sh && chmod +x ~/freespace.sh`.

```sh
#!/bin/bash
set -e

reqSpace=100000000 # 100GB
SPACE=`df "$HOME/torrents" | awk 'END{print $4}'`
if [[ $SPACE -le reqSpace ]]
then
  #echo "not enough space"
  #echo "free $SPACE"
  exit 1
fi
#echo "got space"
#echo "free $SPACE"
exit 0
```

Then in the filter use it in **External script** with full path like `/home/USER/freespace.sh` no args and **expect 0**.

Freespace in that is set to 100GB in `$HOME/torrents`, change as you wish. If there is free space it will exit with 0 and filter will take that as an OK. If less than 100GB free it will exit with 1 and then not match the filter and stop there.

Try that as an option until we make something better integrated.

**Comment by CarlitoGil** [comment]
Hi @zze0s!
This made me have to learn Shell scripting yesterday... 'cause I ain't want to remember VBScript. So thanks.

The script provided above has 2 issues, at least in my tests.
- If the path is wrong or `df` can't access it, `awk` outputs a string that's higher than `reqSpace` and exits with `0`.
- `#!/bin/bash` may exit with `0` executed from a Filter, running Docker on Windows 10. `#!/bin/sh` works.

I expanded the script. Added more exit codes to troubleshoot with the logs.
It takes arguments so it may be customized for each Filter.

### Optional Arguments:

1. Size of Torrent in `bytes`. Use `{{.Size}}` or `0` to ignore.
2. Keep Free Space. Use `GiB` integer or `0` for default.
3. Save Path. Leave `empty` for default.

Example:
_**Filter>External>Script**_
`COMMAND: /scripts/freespace.sh`
`ARGUMENTS: {{.Size}} 100 /torrents`
`EXPECTED EXIT STATUS: 0`

That will keep 100 GiB free after downloading the torrent.
If the feed doesn't provide the size of the torrent and the Filter doesn't check it, then it will be zero.
If torrents are added too quickly one after the other the space might still be shown as free.

Note: I've tested only on Docker, and mounted my `/scripts` and `/torrents` folders.

### Customize the Code
`/torrents` is the default save path, change it to what works for you.
`100` GiB is the default free space to keep.
If the size is not more than **100 KiB** it overrides it with **5 GiB**, in case of a feed parsing error.

### _Script:_ freespace.sh
```shell
#!/bin/sh
set -e

# argument 1: torrent payload size; bytes.
if ! [[ "$1" =~ ^[0-9]*$ ]]; then
  exit 5 # invalid 'size' argument.
fi
if [[ $(($1)) -gt 0 ]]; then
  if [[ $1 -gt 102400 ]]; then
    size=$(($1/1024))
  else # if the size is too small it may not be for the payload.
    size=$((5*1024*1024)) # use larger size.
  fi
else
  size=0 # ignore torrent size.
fi

# argument 2: space to keep free; GiB, integer.
if ! [[ "$2" =~ ^[0-9]*$ ]]; then
  exit 4 # invalid 'keep' argument.
fi
if [[ $(($2)) -gt 0 ]]; then
  keep=$2
else
  keep=100 # default, change as needed.
fi

# argument 3: torrent save path.
if [[ -z "$3" ]]; then
  path="/torrents" # default, change as needed.
else
  path=$3
fi
if ! [[ -d "$path" ]]; then
  exit 3 # invalid 'path' argument.
fi

# get free space available.
have=$((`df "$path" | awk 'END{print $4}'`))

# get minimum free space required.
need=$((($keep*1024*1024)+$size))

# check if the needed free space is available.
if [[ $have -eq 0 ]]; then
  exit 2 # no free space.
elif [[ $have -lt $need ]]; then
  exit 1 # not enough free space.
else
  exit 0 # free space available.
fi
```

Maybe this helps those who don't need it to work remotely.

By the way, I'm loving the app more with every update.

Regards.

**Comment by luckylittle** [comment]
There is an external solution with notification(s):
https://github.com/s0up4200/sizechecker
